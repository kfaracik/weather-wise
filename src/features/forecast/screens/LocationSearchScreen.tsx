import React, {useCallback, useState} from 'react';
import {
  ActivityIndicator,
  ImageBackground,
  SafeAreaView,
  StyleSheet,
  Text,
  View,
  KeyboardAvoidingView,
  Platform,
  Keyboard,
} from 'react-native';
import backgroundImageSource from '@shared/assets/images/bg.webp';
import {Divider, TextInput, TouchableRipple} from 'react-native-paper';
import {LocationForecast} from '../components';
import {debounce} from 'lodash';
import {
  type CityName,
  fetchLocation,
  fetchWeatherForecast,
  Location,
} from '../api';
import {WeatherResponse} from '../api/forecastService';

export const LocationSearchScreen = () => {
  const [locations, setLocations] = useState<Location[]>([]);
  const [inputFocus, setInputFocus] = useState(false);
  const [inputTextValue, setInputTextValue] = useState('');
  const [selectedLocationForecast, setSelectedLocationForecast] =
    useState<WeatherResponse>();
  const [loading, setLoading] = useState(false);

  const handleLocation = async (location: Location) => {
    setLoading(true);
    try {
      const data = await fetchWeatherForecast({
        lat: location.lat,
        lon: location.lon,
      });
      setSelectedLocationForecast(data);
    } catch (error) {
      console.error('Error fetching weather forecast:', error);
    } finally {
      setLoading(false);
    }
    setInputFocus(false);
    setInputTextValue('');
    Keyboard.dismiss();
  };

  const handleLocationSearch = async (cityName: CityName) => {
    setLoading(true);
    try {
      const data = await fetchLocation({cityName});
      setLocations(data);
    } catch (error) {
      console.error('Error fetching location data:', error);
    } finally {
      setLoading(false);
    }
  };

  const handleTextDebounce = useCallback(
    debounce(handleLocationSearch, 1500),
    [],
  );

  const onChangeText = (cityName: CityName) => {
    handleTextDebounce(cityName);
    setInputTextValue(cityName);
  };

  return (
    <ImageBackground
      blurRadius={16}
      source={backgroundImageSource}
      style={styles.backgroundImage}>
      <SafeAreaView style={styles.safeArea}>
        <KeyboardAvoidingView
          behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
          style={styles.container}>
          <View style={styles.contentContainer}>
            <View style={styles.searchContainer}>
              <TextInput
                mode="outlined"
                placeholder="Search city"
                value={inputTextValue}
                outlineStyle={styles.searchInput}
                onChangeText={onChangeText}
                onFocus={() => setInputFocus(true)}
                onBlur={() => setInputFocus(false)}
                style={styles.textInput}
              />
              {loading ? (
                <ActivityIndicator size="large" color="#0000ff" />
              ) : locations?.length > 0 && inputFocus ? (
                <View style={styles.resultsContainer}>
                  {locations.map((location, id) => (
                    <View key={id}>
                      <TouchableRipple
                        style={styles.itemStyle}
                        onPress={() => handleLocation(location)}>
                        <Text style={styles.text}>
                          {location.name}, {location.state}
                        </Text>
                      </TouchableRipple>
                      {id < locations?.length - 1 && <Divider />}
                    </View>
                  ))}
                </View>
              ) : null}
            </View>
            {selectedLocationForecast ? (
              <LocationForecast locationForecast={selectedLocationForecast} />
            ) : !loading ? (
              <Text style={styles.text}>Check the weather for a city</Text>
            ) : null}
          </View>
        </KeyboardAvoidingView>
      </SafeAreaView>
    </ImageBackground>
  );
};

const styles = StyleSheet.create({
  backgroundImage: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
    height: '100%',
  },
  safeArea: {
    flex: 1,
    width: '100%',
  },
  container: {
    flex: 1,
  },
  text: {
    fontSize: 16,
  },
  contentContainer: {
    flex: 1,
    margin: 8,
    gap: 16,
    alignItems: 'center',
  },
  searchContainer: {
    zIndex: 1,
    width: '100%',
  },
  searchInput: {
    borderRadius: 16,
  },
  textInput: {
    opacity: 0.7,
  },
  resultsContainer: {
    marginTop: 4,
    borderRadius: 16,
    backgroundColor: 'white',
  },
  itemStyle: {
    flexDirection: 'row',
    paddingHorizontal: 8,
    paddingVertical: 8,
  },
});
