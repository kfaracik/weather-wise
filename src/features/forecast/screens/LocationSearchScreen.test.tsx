import React from 'react';
import {render, fireEvent, waitFor} from '@testing-library/react-native';
import {LocationSearchScreen} from './LocationSearchScreen';

describe('LocationSearchScreen', () => {
  it('should generate snapshot', () => {
    const {toJSON} = render(<LocationSearchScreen />);
    expect(toJSON()).toMatchSnapshot();
  });
});
