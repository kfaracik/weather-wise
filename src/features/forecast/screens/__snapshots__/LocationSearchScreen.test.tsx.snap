// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationSearchScreen should generate snapshot 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "height": 1334,
      "justifyContent": "center",
      "width": 750,
    }
  }
>
  <Image
    blurRadius={16}
    source={
      {
        "testUri": "../../../src/shared/assets/images/bg.webp",
      }
    }
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "height": 1334,
          "width": 750,
        },
        undefined,
      ]
    }
  />
  <RCTSafeAreaView
    style={
      [
        {
          "flex": 1,
          "margin": 16,
          "width": "100%",
        },
        undefined,
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "gap": 16,
          "margin": 8,
        }
      }
    >
      <View
        style={
          {
            "width": "100%",
            "zIndex": 1,
          }
        }
      >
        <View
          style={
            {
              "opacity": 0.7,
            }
          }
        >
          <View
            pointerEvents="none"
            style={
              [
                {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 6,
                },
                {
                  "top": 0,
                },
                {
                  "backgroundColor": "rgba(255, 251, 254, 1)",
                  "borderColor": "rgba(121, 116, 126, 1)",
                  "borderRadius": 4,
                  "borderWidth": 1,
                },
                {
                  "borderRadius": 16,
                },
              ]
            }
            testID="text-input-outline"
          />
          <View
            style={
              [
                {
                  "flexGrow": 1,
                  "paddingBottom": 0,
                },
                {
                  "minHeight": 56,
                  "paddingTop": 0,
                },
              ]
            }
          >
            <TextInput
              cursorColor="rgba(103, 80, 164, 1)"
              editable={true}
              maxFontSizeMultiplier={1.5}
              multiline={false}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              onLayout={[Function]}
              placeholder="Search city"
              placeholderTextColor="rgba(73, 69, 79, 1)"
              selectionColor="rgba(103, 80, 164, 1)"
              style={
                [
                  {
                    "flexGrow": 1,
                    "margin": 0,
                  },
                  {
                    "height": 56,
                  },
                  {
                    "paddingBottom": 0,
                    "paddingTop": 0,
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "letterSpacing": 0.15,
                    "lineHeight": undefined,
                    "minWidth": 65,
                    "paddingHorizontal": 16,
                    "textAlign": "left",
                    "textAlignVertical": "center",
                  },
                  false,
                  [
                    {},
                  ],
                  undefined,
                ]
              }
              testID="text-input-outlined"
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
        </View>
      </View>
      <Text
        style={
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "fontFamily": "System",
              "fontWeight": "400",
              "letterSpacing": 0,
            },
            {
              "writingDirection": "ltr",
            },
            {
              "color": "lightgray",
              "fontSize": 16,
            },
          ]
        }
      >
        Check the weather for a city
      </Text>
    </View>
  </RCTSafeAreaView>
</View>
`;
